engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java

```diff
@@ -53,204 +56,36 @@ public abstract Stateful execute(
-      Object[] arguments,
-      CallArgumentInfo[] schema,
-      InvokeCallableNode.DefaultsExecutionMode defaultsExecutionMode,
-      InvokeCallableNode.ArgumentsExecutionMode argumentsExecutionMode,
-      BaseNode.TailStatus isTail,
-      @Cached BooleanResolverNode booleanResolverNode,
-      @Cached IndirectInvokeFunctionNode invokeFunctionNode) {
-    Function function = booleanResolverNode.execute(symbol, _this);
-    return invokeFunctionNode.execute(
-        function,
-        frame,
-        state,
-        arguments,
-        schema,
-        defaultsExecutionMode,
-        argumentsExecutionMode,
-        isTail);
-  }
-
-  @Specialization
-  Stateful doText(
+  @Specialization(guards = "dispatch.hasFunctionalDispatch(_this)")
+  Stateful doFunctionalDispatch(
       MaterializedFrame frame,
       Object state,
       UnresolvedSymbol symbol,
-      Text _this,
-      Object[] arguments,
-      CallArgumentInfo[] schema,
-      InvokeCallableNode.DefaultsExecutionMode defaultsExecutionMode,
-      InvokeCallableNode.ArgumentsExecutionMode argumentsExecutionMode,
-      BaseNode.TailStatus isTail,
-      @Cached TextResolverNode textResolverNode,
-      @Cached IndirectInvokeFunctionNode invokeFunctionNode) {
-    Function function = textResolverNode.execute(symbol, _this);
-    return invokeFunctionNode.execute(
-        function,
-        frame,
-        state,
-        arguments,
-        schema,
-        defaultsExecutionMode,
-        argumentsExecutionMode,
-        isTail);
-  }
-
-  @Specialization
-  Stateful doFunction(
-      MaterializedFrame frame,
-      Object state,
-      UnresolvedSymbol symbol,
-      Function _this,
+      Object _this,
       Object[] arguments,
       CallArgumentInfo[] schema,
       InvokeCallableNode.DefaultsExecutionMode defaultsExecutionMode,
       InvokeCallableNode.ArgumentsExecutionMode argumentsExecutionMode,
       BaseNode.TailStatus isTail,
-      @Cached FunctionResolverNode functionResolverNode,
-      @Cached IndirectInvokeFunctionNode invokeFunctionNode) {
-    Function function = functionResolverNode.execute(symbol, _this);
-    return invokeFunctionNode.execute(
-        function,
-        frame,
-        state,
-        arguments,
-        schema,
-        defaultsExecutionMode,
-        argumentsExecutionMode,
-        isTail);
+      @CachedLibrary(limit = "10") MethodDispatchLibrary dispatch,
+      @Cached IndirectInvokeFunctionNode invokeFunctionNode,
+      @CachedContext(Language.class) TruffleLanguage.ContextReference<Context> ctx) {
+    try {
+      Function function = dispatch.getFunctionalDispatch(_this, symbol);
+      return invokeFunctionNode.execute(
+          function,
+          frame,
+          state,
+          arguments,
+          schema,
+          defaultsExecutionMode,
+          argumentsExecutionMode,
+          isTail);
+    } catch (MethodDispatchLibrary.NoSuchMethodException e) {
+      CompilerDirectives.transferToInterpreter();
```

Comment:
We really need a consistent rule for whether we use this or not when throwing.

Reply:
Fair. Let's stick with `don't use` for new code.